lib/src/bootloader.rs

```diff
@@ -26,3 +26,113 @@ pub(crate) fn install_via_bootupd(
         .verbose()
         .run()
 }
+
+#[context("Installing bootloader using zipl")]
+pub(crate) fn install_via_zipl(
+    device: &Utf8Path,
+    boot_uuid: &str,
+) -> Result<()> {
+
+    // Run findmnt to find the boot partition
+
+    #[derive(serde::Deserialize)]
+    struct Filesystem {
+        source: String,
+        target: String,
+    }
+
+    #[derive(serde::Deserialize)]
+    struct Findmnt {
+        filesystems: Vec<Filesystem>,
+    }
+
+    let findmnt_desc = format!("Running findmnt to identify the boot partition with UUID={boot_uuid}");
+    let findmnt_task = Task::new(findmnt_desc, "findmnt");
+    let findmnt_args = ["--json", "--output", "TARGET,SOURCE,FSTYPE",
+                        "--uniq", "--first-only", "--nofsroot",
+                        "--source", &(format!("UUID={boot_uuid}"))];
+    let findmnt_out  = findmnt_task.args(findmnt_args).quiet().read()?;
+    let findmnt_json = serde_json::from_str::<Findmnt>(&findmnt_out).context("parsing findmnt output")?;
+
+    let boot_fs = findmnt_json.filesystems.get(0).context("finding boot filesystem")?;
+    let boot_dev = Utf8Path::new(&boot_fs.source);
+    let boot_dir = Utf8Path::new(&boot_fs.target);
+
+    // Run sfdisk to determine the boot partition offset
+
+    #[derive(serde::Deserialize)]
+    struct Partition {
+        node:  String,
+        start: i64,
+    }
+
+    #[derive(serde::Deserialize)]
+    struct PartitionTable {
+        sectorsize: i64,
+        partitions: Vec<Partition>,
+    }
+
+    #[derive(serde::Deserialize)]
+    struct Sfdisk {
+        partitiontable: PartitionTable
+    }
+
+    let sfdisk_desc = format!("running sfdisk to determine the partition offset of {boot_dev}");
+    let sfdisk_task = Task::new(sfdisk_desc, "sfdisk");
+    let sfdisk_args = ["--json", device.as_str()];
+    let sfdisk_out  = sfdisk_task.args(sfdisk_args).quiet().read()?;
+    let sfdisk_json = serde_json::from_str::<Sfdisk>(&sfdisk_out).context("parsing findmnt output")?;
+
+    let block_size  = sfdisk_json.partitiontable.sectorsize;
+    let partitions  = sfdisk_json.partitiontable.partitions;
+    let boot_part   = partitions.iter().find(|p| p.node == boot_dev).context("finding boot partition")?;
+    let boot_offset = boot_part.start;
+
+    // Find the default BLS conf under /boot/loader/entries
+
+    let bls_dir = boot_dir.join("boot/loader/entries");
+
+    let ls_desc = format!("running ls to find the default BLS configuration");
+    let ls_task = Task::new(ls_desc, "ls");
+    let ls_args = ["--sort=version", "--reverse", bls_dir.as_str()];
+    let ls_out = ls_task.args(ls_args).quiet().read()?;
+
+    let bls_name = ls_out.lines().find(|filename| filename.ends_with(".conf")).context("finding BLS conf")?;
+    let bls_path = bls_dir.join(bls_name);
+    let bls_conf = std::fs::read_to_string(&bls_path).context(format!("reading {bls_path:?}"))?;
+
+    let mut image = "";
```

Comment:
I prefer the coreos-installer version of this where we use `Option<T>`: https://github.com/coreos/coreos-installer/blob/803b911dabac28b958a89a6b3ee01ce2df65f99a/src/s390x/zipl.rs#L202 versus representing "unset" as the empty string.

Reply:
OK. I'll fix the code.